core/planner/planner.go

```diff
@@ -55,9 +55,13 @@ func (t *Planner) plan(rates api.Rates, requiredDuration time.Duration, targetTi
 		slotDuration := slot.End.Sub(slot.Start)
 		requiredDuration -= slotDuration
 
-		// slot covers more than we need, so lets start late
+		// slot covers more than we need, so shorten it
 		if requiredDuration < 0 {
-			slot.Start = slot.Start.Add(-requiredDuration)
+			if SlotHasPredecessor(slot, plan) || !SlotHasSuccessor(slot, plan) {
+				slot.End = slot.End.Add(requiredDuration)
```

Comment:
Also grundsätzlich finde ich die Änderung super.

Aber eine Frage: Ist das hier wirklich sinnvoll? Also wenn man z.B. nur noch 10min laden muss, aber der Slot 60 min ist, ziehen wir hier 50min ab. Aber wie wahrscheinlich ist, dass das stimmt? Kann man den Slot nicht in dem Fall einfach auf 60min lassen und wenn er fertig ist hört er auf? Also für den Fall das er vorne startet ...

Ich würde also das IF genau umdrehen und nur den folgenden Fall anpassen: slot.Start = slot.Start.Add(-requiredDuration) Also grob wie vorher nur mit mehr Bedingungen im IF

Reply:
Dann vermischen wir ja wieder die Aufgaben. Aufgabe Planner: Plan für Zeit entwickeln, das tut er. Aufgabe Loadpoint: Plan anwenden und ggf. Justieren. Es gibt ja ständig einen neuen Plan- wenn der LP sieht das es länger dauert dann verschieben sich ja hier auch die Grenzen. So wie der Code jetzt ist sorgen wir dafür, dass wir meist am Anfang starten und nur ausgewählten Fällen (nämlich wenn danach noch was kommt und vorher nix, z.B. wenn der teuerste Slot der zeitlich allererste ist) am Ende. Das sollte die wenigstens Lücken produzieren.

Falls ich dich falsch verstanden habe: gerne konkretes Beispiel!

Reply:
Ok ich probiere mich noch mal.
1. Wir schränken jetzt folgendes ein mit der Änderung: slot.Start = slot.Start.Add(-requiredDuration), das halte ich für richtig.
2. Frage:
Ich verstehe die Änderung so.
- Wenn der Slot länger ist als wir ihn brauchen (was implizit heisst wir sind im letzten slot)
- Machen wir ihn kürzer.
- In dem neuen Beispiel starten wir unter gewissen umständen von vorne (was ich gut finde)

Genau in dem Fall Frage ich mich, welchen mehrwert es hat ihn zu kürzen. Wenn es der letzte Slot ist, ist es doch egal ob er gekürzt wird oder? Also welchen mehrwert hat es?

Reply:
> Genau in dem Fall Frage ich mich, welchen mehrwert es hat ihn zu kürzen

Aufgabe des Planners ist einen Plan definierter Länge zu erzeugen. Wir können das gerne später aufweichen wenn wir sehen, dass es ein Problem damit gibt. Vorher hätte ichs gerne "richtig".

Reply:
Der Loadpoint stellt sicher dass nach dem letzten Slot nicht beendet wird falls das Ladeziel nicht erreicht ist. Da der zeitlich letzte Ladeslot keinen nachfolgenden hat fängt der also immer "früh" an. Für mich passt das erstmal?!

Reply:
Ok grundsätzlich geht deine Änderung auch und ist ja mathematisch richtig. Ich würde es anders programmieren und wollte es nur vorschlagen. Also ok von meiner Seite.

Reply:
Mir kommt grad noch ne ganz andere Idee. Sollten wir nicht einfach nur den zeitlich ersten Slot spät starten?

Nur dieser könnte ja noch von PV profitieren. Alle anderen Slots sollten immer früh starten. D.h. *nur* und nur dann wenn der zeitlich erste Slot der teuerste ist würden wir später beginnen.

Sollte sich da im Ladeverlauf ändern, also wenn es Verschiebungen gäbe und der teuerste Slot plötzlich in der Mitte wäre dann müsste man ja einfach nur die Zeit abwarten bis dieser Slot dran wäre. Dann gäbs einen neuen Plan und er wäre wieder der erste.

Was meinst Du?

Reply:
Der aktuelle Code ist eh noch falsch weil die Funktionen nur funktionieren, wenn der Plan nach Zeit sortiert ist- das sind sie in `plan()` aber gar nicht (facepalm).

Reply:
> Mir kommt grad noch ne ganz andere Idee. Sollten wir nicht einfach nur den zeitlich ersten Slot spät starten?
> 
> Nur dieser könnte ja noch von PV profitieren. Alle anderen Slots sollten immer früh starten. D.h. _nur_ und nur dann wenn der zeitlich erste Slot der teuerste ist würden wir später beginnen.
> 
> Sollte sich da im Ladeverlauf ändern, also wenn es Verschiebungen gäbe und der teuerste Slot plötzlich in der Mitte wäre dann müsste man ja einfach nur die Zeit abwarten bis dieser Slot dran wäre. Dann gäbs einen neuen Plan und er wäre wieder der erste.
> 
> Damit würde sich die Frage in `IsFirst` ändern.
> 
> Was meinst Du?

Also ich finde es gut und einfach.
Du kannst natürlich auch preislich theoretisch 3 gestückelte slots haben. Also 1 Stunde laden, 1 Stunde pause, 1 Stunde laden, eine Stunde pause, eine Stunde laden. Dann müssten theoretisch alle 3 slots spät starten oder?

Reply:
Das wär so

Reply:
Jetzt muss ich doof fragen. In meinem Beispiel ist der letzte Slot in dem Moment wo man da ankommt halt auch der letzte. Da er aber keinen Nachfolger hat würden wir ihn trotzdem vorne Starten wollen. Klappt das mit der First abfrage oder nicht?

Reply:
Oder brauchen wir einen ist first und nicht der letzte? Dann sind wir aber fast wieder beim ersten Vorschlag von dir.

Reply:
Neuer Test sollte das bestätigen:

	// ensure single slot is always first
	require.True(t, IsFirst(first, []api.Rate{first}))

Reply:
Good catch, da hab ich mir eben beim Testen auch den Kopf über den Testcase zerbrochen. Dann ändern wir das gleich mal:

	// the first (if not single) slot should start as late as possible
	if IsFirst(slot, plan) && len(plan) > 0 {
		slot.Start = slot.Start.Add(-requiredDuration)
	} else {
		slot.End = slot.End.Add(requiredDuration)
	}

Reply:
So, jetzt nochmal alle Testcases überarbeitet und auch randomisiert und Fehler behoben.